lib/Polaris.Class.ps1

```diff
@@ -2,25 +2,140 @@ class Polaris {
     
     [Int]$Port
     [System.Collections.Generic.List[PolarisMiddleWare]]$RouteMiddleWare = [System.Collections.Generic.List[PolarisMiddleWare]]::new()
-    [System.Collections.Generic.Dictionary[string, [System.Collections.Generic.Dictionary[string, string]]]]$ScriptBlockRoutes = [System.Collections.Generic.Dictionary[string, [System.Collections.Generic.Dictionary[string, string]]]]::new()
+    [System.Collections.Generic.Dictionary[string, [System.Collections.Generic.Dictionary[string, scriptblock]]]]$ScriptBlockRoutes = [System.Collections.Generic.Dictionary[string, [System.Collections.Generic.Dictionary[string, scriptblock]]]]::new()
     hidden [Action[string]]$Logger
     hidden [System.Net.HttpListener]$Listener
-    hidden [System.Management.Automation.Runspaces.RunspacePool]$PowerShellPool = [RunspaceFactory]::CreateRunspacePool()
     hidden [bool]$StopServer = $False
     [string]$getLogsString = "PolarisLogs"
     [string] $ClassDefinitions = $script:ClassDefinitions
+    $ContextHandler = (New-ScriptBlockCallback -Callback {
+
+            param(
+                [System.IAsyncResult]
+                $AsyncResult
+            ) 
+
+            [Polaris]$Polaris = $AsyncResult.AsyncState
+            $context = $Polaris.Listener.EndGetContext($AsyncResult)
+        
+
+            if ($Polaris.StopServer -or $null -eq $context) {
+                if ($null -ne $Polaris.Listener) {
+                    $Polaris.Listener.Close()
+                }
+                break
+            }
+
+            $Polaris.Listener.BeginGetContext($Polaris.ContextHandler, $Polaris)
+
+            [System.Net.HttpListenerRequest] $rawRequest = $context.Request
+            [System.Net.HttpListenerResponse] $rawResponse = $context.Response
+
+            $Polaris.Log("request came in: " + $rawRequest.HttpMethod + " " + $rawRequest.RawUrl)
+
+            [PolarisRequest] $request = [PolarisRequest]::new($rawRequest)
+            [PolarisResponse] $response = [PolarisResponse]::new()
+
+            
+            [string] $route = $rawRequest.Url.AbsolutePath.TrimEnd('/')
+
+            if ( [string]::IsNullOrEmpty($route) ) { $route = "/" }
+
+            try {
+
+                # Run middleware in the order in which it was added
+                foreach ($middleware in $Polaris.RouteMiddleware) {
+                    $ScriptBlock = [scriptblock]::Create(
+                        "param(`$request,`$Response)`r`n" +
+                        $middleware.ScriptBlock.ToString()
+                    )
+                    
+                    Invoke-Command -ScriptBlock $ScriptBlock `
+                        -ArgumentList @($Request, $Response)
+                }
+
+                $Polaris.Log("Parsed Route: $Route")
+                $Polaris.Log("Request Method: $($rawRequest.HttpMethod)")
+                $Routes = $Polaris.ScriptBlockRoutes
+                $MatchingRoute = $Routes.keys.where( {$route -match $_})[0]
```

Comment:
Isn't this some form of dictionary and can just be:
```powershell
if ( $Routes[$route] -and
    $Routes[$route][$Method] ) {
    
    # do invoke command
}
```

it's an O(1) lookup vs O(n)

Reply:
I see what you're saying here and I went to put it in but then I remembered how / why I put that in. PowerShell  will throw an index evaluates to null exception if the route doesn't match but it's not one that I could catch for some reason, it would just register the error and then continue on without moving to the catch block. Not sure if you know of some magic there but we may want to register this as an issue for later.

The other thing to consider would be that I believe we would need to move the static routes back from serving all files in a directory to being a list of paths added when the function is called which means new files added to the directory after the server started wouldn't get added unless the function was called again.